test/parse/test_entity_declaration.rb

```diff
@@ -23,6 +23,158 @@ def parse(internal_subset)
+          assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 60
+Last 80 unconsumed characters:
+ foo SYSTEM \"bar\" xxx NDATA gif > ]>
+          DETAIL
+        end
+
+        def test_garbaged_after_ndata_declaration
+          exception = assert_raise(REXML::ParseException) do
+            REXML::Document.new('<!DOCTYPE root [<!ENTITY foo SYSTEM "bar" NDATA gif xxx > ]>')
+          end
+          assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 60
+Last 80 unconsumed characters:
+ foo SYSTEM \"bar\" NDATA gif xxx > ]>
+          DETAIL
+        end
+      end
+
+      class TestPublic < self
+        def test_garbaged_between_public_and_first_literal
+          exception = assert_raise(REXML::ParseException) do
+            REXML::Document.new('<!DOCTYPE root [<!ENTITY foo PUBLIC xxx "bar" "baz" > ]>')
+          end
+          assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 56
+Last 80 unconsumed characters:
+ foo PUBLIC xxx \"bar\" \"baz\" > ]>
+          DETAIL
+        end
+
+        def test_garbaged_between_literals
+          exception = assert_raise(REXML::ParseException) do
+            REXML::Document.new('<!DOCTYPE root [<!ENTITY foo PUBLIC "bar" xxx "baz" > ]>')
+          end
+          assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 56
+Last 80 unconsumed characters:
+ foo PUBLIC \"bar\" xxx \"baz\" > ]>
+          DETAIL
+        end
+
+        def test_garbaged_after_literals
+          exception = assert_raise(REXML::ParseException) do
+            REXML::Document.new('<!DOCTYPE root [<!ENTITY foo PUBLIC "bar" "baz" xxx > ]>')
+          end
+          assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 56
+Last 80 unconsumed characters:
+ foo PUBLIC \"bar\" \"baz\" xxx > ]>
+          DETAIL
+        end
+
+        def test_garbaged_between_literal_and_ndata_declaration
+          exception = assert_raise(REXML::ParseException) do
+            REXML::Document.new('<!DOCTYPE root [<!ENTITY foo PUBLIC "bar" "baz" xxx NDATA gif > ]>')
+          end
+          assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 66
+Last 80 unconsumed characters:
+ foo PUBLIC \"bar\" \"baz\" xxx NDATA gif > ]>
+          DETAIL
+        end
+
+        def test_garbaged_after_ndata_declaration
+          exception = assert_raise(REXML::ParseException) do
+            REXML::Document.new('<!DOCTYPE root [<!ENTITY foo PUBLIC "bar" "baz" NDATA gif xxx > ]>')
+          end
+          assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 66
+Last 80 unconsumed characters:
+ foo PUBLIC \"bar\" \"baz\" NDATA gif xxx > ]>
+          DETAIL
+        end
+      end
+    end
+
+    class TestCharacterCode < self
+      def test_pubid_literal
```

Comment:
```suggestion
    class TestPublicIDLiteral < self
      def test_invalid_pubid_char
```

Could you add `PubidLiteral` and `PubidChar` definitions as a comment to understand why this value is invalid?

* https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-PubidLiteral
* https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-PubidChar